From 9094c9a14ccba6517863c939e017eb824211fef3 Mon Sep 17 00:00:00 2001
From: Elaine Zhang <zhangqing@rock-chips.com>
Date: Mon, 27 Apr 2020 09:39:55 +0800
Subject: [PATCH] clk: rockchip: Modify the rules of the fractional divider

Fractional frequency divider to do integer frequency divider does not
need 20 times the limit.

Change-Id: I98e413230a237fc5aafebabb9c0696d0ce51281a
Signed-off-by: Elaine Zhang <zhangqing@rock-chips.com>
---
 drivers/clk/rockchip/clk.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/clk/rockchip/clk.c b/drivers/clk/rockchip/clk.c
index 60533e8caa3f..049cd405c7a2 100644
--- a/drivers/clk/rockchip/clk.c
+++ b/drivers/clk/rockchip/clk.c
@@ -213,6 +213,16 @@ static void rockchip_fractional_approximation(struct clk_hw *hw,
 		}
 
 		if (*parent_rate < rate * 20) {
+			/*
+			 * Fractional frequency divider to do
+			 * integer frequency divider does not
+			 * need 20 times the limit.
+			 */
+			if (!(*parent_rate % rate)) {
+				*m = 1;
+				*n = *parent_rate / rate;
+				return;
+			}
 			pr_warn("%s p_rate(%ld) is low than rate(%ld)*20, use integer or half-div\n",
 				clk_hw_get_name(hw), *parent_rate, rate);
 			*m = 0;
-- 
2.35.3

